### PR TITLE
feat(computed): returns previous value in case of error

### PR DESCRIPTION
**Pull request by HamadaBrest**
fixes https://github.com/JsCommunity/reaclette/issues/21

### CODE REVIEW

File: src/factory.js [hunk 1]
```diff
@@ -137,7 +137,13 @@ module.exports = ({ Component, createElement, PropTypes }) => {
 
                   propsSpy.clear()
                   stateSpy.clear()
-                  previousValue = c(stateProxy, propsProxy)
+
+                  try {
+                    previousValue = c(stateProxy, propsProxy)
+                  } catch (error) {
+                    noop()
```
Comment (julien-f) [inline src/factory.js]: What does it do?

**Comment by julien-f** [comment]
Please add some unit tests for this :slightly_smiling_face:

File: src/factory.js (hunk 1 above)
Comment (HamadaBrest) [reply src/factory.js]: Nothing. According to my tests, when I wrap `previousValue = c(stateProxy, propsProxy)` with `try and catch` it don't trigger a render and still expose the previous value in subsequent render as it descriped in issue https://github.com/JsCommunity/reaclette/issues/21
So I added `noop()` to say that when we catch an error, we do nothing.

Comment (julien-f) [reply src/factory.js]: Do not call a function if there is nothing to do.

It's better to add a comment explaining why you are swallowing the error.

**Comment by julien-f** [comment]
Please rebase this PR and properly handle `CircularComputedError`.

**Comment by HamadaBrest** [comment]
> Please rebase this PR and properly handle `CircularComputedError`.

I'll do it as fast as possible :1st_place_medal:

File: src/factory.js [hunk 2]
```diff
@@ -149,7 +149,10 @@ module.exports = ({ Component, createElement, PropTypes }) => {
                   try {
                     previousValue = c(stateProxy, propsProxy)
                   } catch (error) {
-                    // The fact that the error is catched here, it doesn't update computed's state, so no rendering will be triggered and the computed keeps its previous state
+                    if (error instanceof CircularComputedError) {
+                      throw error
+                    }
+                    // The fact that the error is catched here, it doesn't update computed state, so no rendering will be triggered and the computed keeps its previous state
```
Comment (julien-f) [inline src/factory.js]: Not very clear, here's my proposal:

```js
// as per #21,  keep the previous value in case of error
```

Comment (julien-f) [inline src/factory.js]: Also, I think we should display the errors in the console to help debug.

Comment (HamadaBrest) [reply src/factory.js]: If ` c(stateProxy, propsProxy)` throws, `previousValue` will be not updated, are you sure that we have to do something in order to keep its value ?

File: src/factory.js [hunk 3]
```diff
@@ -145,7 +145,17 @@ module.exports = ({ Component, createElement, PropTypes }) => {
...
                   stateSpy.clear()
-                  previousValue = c(stateProxy, propsProxy)
+
+                  try {
+                    previousValue = c(stateProxy, propsProxy)
+                  } catch (error) {
+                    if (error instanceof CircularComputedError) {
+                      throw error
+                    }
+                    console.log(`computed "${k}" throwed`, error)
```
Comment (julien-f) [inline src/factory.js]: [`console.warn`](https://devdocs.io/dom/console/warn) would be more appropriate.

Comment (julien-f) [inline src/factory.js]: *throwed* is incorrect, it's *thrown*.

File: src/index.spec.js [hunk 4]
```diff
@@ -206,6 +210,13 @@ describe('provideState', () => {
       }).toThrowError(new CircularComputedError('circularComputed'))
     })
 
+    it('does not trigger a render when a computed throws', () => {
```
Comment (julien-f) [inline src/index.spec.js]: I don't believe this is what you're testing here.

Comment (julien-f) [reply src/index.spec.js]: What you need to test:
- a computed that throws on the first call is visible as `undefined` in the state
- a computed that returns a value and then throws is visible as the previous value

File: src/index.spec.js [hunk 5]
```diff
@@ -161,11 +161,18 @@ describe('provideState', () => {
   describe('computed', () => {
     const sum = jest.fn(({ foo }, { bar }) => foo + bar)
     const circularComputed = jest.fn(({ circularComputed }) => {})
+    const throwComputed = jest.fn((_, { baz }) => {
```
Comment (julien-f) [inline src/index.spec.js]: Unsupported: `baz` is not declared in `initialState`.

File: src/index.spec.js [hunk 6]
```diff
@@ -206,6 +213,19 @@ describe('provideState', () => {
...
+      setParentProps({ baz: 21 })
+      const res = getInjectedState().throwComputed
+      expect(res).toBe(undefined)
+
+      // a computed that returns a value and then throws is visible as the previous value
+      setParentProps({ baz: 5 })
+      expect(getInjectedState().throwComputed).toBe(10)
+      setParentProps({ baz: 21 })
+      expect(getInjectedState().throwComputed).toBe(10)
+    })
```
Comment (julien-f) [inline src/index.spec.js]: Why not split this in two tests?

File: src/index.spec.js (hunk 5 above)
Comment (HamadaBrest) [reply src/index.spec.js]: I know, computed can be derived from props too. here `baz` is a prop.. Should we test only computed derived from State ?

Comment (julien-f) [reply src/index.spec.js]: My bad, this is a `prop`, this is correct.

**Pull request merged**